src/Sentry/Internal/BackgroundWorker.cs

```diff
@@ -156,6 +166,76 @@ public bool EnqueueEvent(SentryEvent @event)
             }
         }
 
+        public async Task FlushAsync(TimeSpan timeout)
+        {
+            if (_disposed)
+            {
+                _options.DiagnosticLogger?.LogDebug("Worker disposed. Nothing to flush.");
+                return;
+            }
+
+            if (!_queue.TryPeek(out _))
+            {
+                _options.DiagnosticLogger?.LogDebug("No events to flush.");
+                return;
+            }
+
+            // Start timer from here.
+            var timeoutSource = new CancellationTokenSource();
+            timeoutSource.CancelAfter(timeout);
+            var flushSuccessSource = new CancellationTokenSource();
+
+            var timeoutWithShutdown = CancellationTokenSource.CreateLinkedTokenSource(
+                timeoutSource.Token,
+                _shutdownSource.Token,
+                flushSuccessSource.Token);
+
+            var objToWait = new object();
+
+            void EventFlushedCallback(object objProcessed, EventArgs _)
+            {
+                if (ReferenceEquals(objToWait, objProcessed))
+                {
+                    try
+                    {
+                        _options.DiagnosticLogger?.LogDebug("Signaling flush completed.");
+                        // ReSharper disable once AccessToDisposedClosure
+                        flushSuccessSource.Cancel();
+                    }
+                    catch // Timeout or Shutdown might have been called so this token was disposed.
+                    {
+                    } // Flush will release when timeout is hit.
+                }
+            }
+
+            OnFlushObjectReceived += EventFlushedCallback;
+            try
+            {
+                // Wait for a slot in the queue or one of the tokens triggers
+                var acquired = await _inSemaphore.WaitAsync(timeout, timeoutWithShutdown.Token).ConfigureAwait(false);
```

Comment:
If you do `WaitAsync(timeout, ..)` here and further down `Task.Delay(timeout, ..)`, doesn't that mean you will potentially wait for `2 * timeout`?

I would calculate a `deadline` timestamp at the start of this function and then calculate a new timeout for the `Task.Delay`

Reply:
No because the `timeoutSource` `CancellationToken` is created based on the timeout.
In this case, once the timeout was reached, either at `AwaitAsync` or at `Task.Delay` an `TaskCanceledException` would be thrown (and caught by the `OperationCanceledException` catch block,

Reply:
Meaning I could set int.MaxValue there on the `WaitAsync` or `Task.Delay` but both combined would never "block" longer than _timeout_